### PR TITLE
Fixed NaN's GetHashCode and Equals invariant issue.

### LINKED ISSUE #6237: double.GetHashCode and double.Equals do not agree (When NaN and -NaN are compared for equality)
It is an invariant that two objects for which `o1.Equals(o2)`, their hash codes should agree, i.e. `o1.GetHashCode() == o2.GetHashCode()`. See [the specification](https://msdn.microsoft.com/en-us/library/system.object.gethashcode%28v=vs.110%29.aspx), which says

> Two objects that are equal return hash codes that are equal.

This invariant is assumed by many APIs that use these methods, and any types that fail to satisfy them are "broken" and cannot be expected to work with the core APIs. However, the implementation for `double` does not satisfy this invariant. Specifically, the following snippet of code prints "True False".

``` cs
        object o1 = double.NaN;
        object o2 = -double.NaN;
        Console.WriteLine(o1.Equals(o2));
        Console.WriteLine(o1.GetHashCode() == o2.GetHashCode());
```

### PR DESCRIPTION
**Pull request by jkotas**
Fix #6237

### CODE REVIEW

File: src/mscorlib/shared/System/Double.cs [hunk 1]
```diff
@@ -221,16 +223,18 @@ public bool Equals(Double obj)
         //The hashcode for a double is the absolute value of the integer representation
         //of that double.
         //
+        [MethodImpl(MethodImplOptions.AggressiveInlining)] // 64-bit constants make the IL unusually large that makes the inliner to reject the method
         public unsafe override int GetHashCode()
         {
-            double d = m_value;
-            if (d == 0)
+            var bits = Unsafe.As<double, long>(ref m_value);
```
Comment (mikedn) [inline src/mscorlib/shared/System/Double.cs]: I guess you could drop "unsafe" from the method def now that `Unsafe.As` is used.

**Comment by jkotas** [comment]
Perf results (source https://gist.github.com/jkotas/ab888c04052dada1648d50b11bc73412):

|             | Virtual double | Virtual float | Non-virtual double | Non-virtual float |
|-------------|----------------|---------------|--------------------|-------------------|
| Baseline    | 3203           | 3187          | 1109               | 969               |
| #13781      | 5172           | 4156          | 1594               | 1297              |
| This change | 3562           | 3250          | 985                | 640               |

**Comment by jkotas** [comment]
I have experimented with different implementations. This was the best one.

There are 4 interesting cases: double vs. float and virtual vs. non-virtual (ie inlined) call of GetHashCode. This change makes the virtual call a bit slower, and non-virtual call quite a bit faster.

**Comment by jkotas** [comment]
Tests in https://github.com/dotnet/corefx/pull/27438

**Comment by mikedn** [comment]
Presumably we're ok with GetHashCode(0) returning a different value than before?

File: src/mscorlib/shared/System/Double.cs [hunk 2]
```diff
@@ -221,16 +223,18 @@ public bool Equals(Double obj)
...
+
+            if (((bits - 1) & 0x7FFFFFFFFFFFFFFF) >= 0x7FF0000000000000)
```
Comment (tannergooding) [inline src/mscorlib/shared/System/Double.cs]: Why `(bits - 1)`?

Comment (tannergooding) [reply src/mscorlib/shared/System/Double.cs]: This definitely needs a comment.

At first glance, it looks like you are grabbing all values matching (IsInfinity || IsNaN), but this is actually doing (IsNaN || IsZero).

Comment (jkotas) [reply src/mscorlib/shared/System/Double.cs]: Comment added

**Comment by jkotas** [comment]
> Presumably we're ok with GetHashCode(0) returning a different value than before?

It still returns 0 like before, for both positive and negative zero.

**Comment by mikedn** [comment]
> It still returns 0 like before, for both positive and negative zero.

Hah, looks like I wasn't completely awake. Something like "well, `(0 - 1) & 0x7F800000` is `0x7F800000` so it's not 0 anymore"...

While I'm at it: I ran some benchmarks involving `Dictionary<float|double, int>`. In one case the double version was 10% slower and the float version was 10% faster (probably the pesky large constants take their toll). Other cases showed no difference.

**Comment by jkotas** [comment]
@dotnet-bot test Windows_NT armlb Cross Checked Innerloop Build and Test please

**Comment by jkotas** [comment]
> While I'm at it: I ran some benchmarks involving

@mikedn Thanks. What was the case that got 10% slower?

**Comment by mikedn** [comment]
> What was the case that got 10% slower?

I re-run the tests, here are the full results:
Before:

Method |      t |     Mean |     Error |    StdDev 
------- |------- |---------|----------|----------
DMap | (0, 0) | 12.26 ns | 0.1133 ns | 0.1005 ns
FMap | (0, 0) | 12.00 ns | 0.1581 ns | 0.1479 ns 
DMap | (1, 1) | 16.64 ns | 0.0661 ns | 0.0619 ns 
FMap | (1, 1) | 16.46 ns | 0.0663 ns | 0.0588 ns
DMap | (2, 2) | 11.05 ns | 0.0243 ns | 0.0203 ns
FMap | (2, 2) | 11.52 ns | 0.0103 ns | 0.0092 ns

After:

Method |      t |     Mean |     Error |    StdDev 
------- |------- |---------|----------|----------
DMap | (0, 0) | 13.68 ns | 0.2705 ns | 0.2778 ns
FMap | (0, 0) | 10.94 ns | 0.0165 ns | 0.0138 ns
DMap | (1, 1) | 17.02 ns | 0.0554 ns | 0.0518 ns
FMap | (1, 1) | 15.38 ns | 0.1055 ns | 0.0986 ns
DMap | (2, 2) | 12.09 ns | 0.0413 ns | 0.0386 ns
FMap | (2, 2) | 10.92 ns | 0.0105 ns | 0.0093 ns

**Pull request merged**